### PR TITLE
Medical citation contexts

### PR DESCRIPTION
**Pull request by jpowerwa**
Update reference extraction logic to handle references from PLOS papers. Also fixes some incorrect reference publication dates and gets rid of some extraneous spaces in reference content. See comments below for details.

Verified that the unit tests pass and that test output for "Missed extraction" matches the same output for master.

https://github.com/allenai/scholar/issues/10629

### CODE REVIEW

**Comment by jpowerwa** [comment]
Summary reference and reference mention counts for test papers found by science-parse, without and with this set of changes:

```
joannap@joannap:/ws/projects/medical-citations$ grep SUMMARY gold/paper-data.gold.txt 
SUMMARY: 0c5cf03abbcaeb86585ef110141b76ef8765ceb9, references: 19, referenceMentions: 23
SUMMARY: 12cd18f35b83d3559c2be283d1ab0d513bac3ed9, references: 19, referenceMentions: 22
SUMMARY: 1925673c0ac47e9024440db9133d5214fc4104a6, references: 6, referenceMentions: 0
SUMMARY: 1dca9541f4b0fc54c68bd58980602f2dc16662d9, references: 1, referenceMentions: 1
SUMMARY: 295a641761198ea565fe7e877912677f61680dc4, references: 0, referenceMentions: 0
SUMMARY: 2cd5827927d5f960b22cd0db843ae4a56953fb89, references: 85, referenceMentions: 106
SUMMARY: 58d74b5812ce9c975145f592848d3ae6204082af, references: 0, referenceMentions: 0
SUMMARY: 676d59b9330e084b2d5a3718233396a8bd2d02a4, references: 41, referenceMentions: 0
SUMMARY: 68a88ae2b8c586a49876ff4d0af871ed72d42b66, references: 54, referenceMentions: 47
SUMMARY: 770509379ba6d456d35734e8dba0c6367e6e26c2, references: 0, referenceMentions: 0
SUMMARY: 7e9537bbead3c32c07496f7feb787677a2e77613, references: 0, referenceMentions: 0
SUMMARY: 830476f35dc4ad59f86ece5028a030b05ab83696, references: 0, referenceMentions: 0
SUMMARY: 9444ba92b6d5d431e7bb3be26a2d6133f38c5485, references: 26, referenceMentions: 25
SUMMARY: 9e257dad7923d5b44b236ea03fd2c9d654c06a25, references: 30, referenceMentions: 30
SUMMARY: 9f77b3c5b0b821acbc45098e036c3cdc306a4716, references: 25, referenceMentions: 38
SUMMARY: b48ab369f5a3515b9b347cff89b713af51b06866, references: 39, referenceMentions: 61
SUMMARY: b5252acfddf1868c6b1a8078d6c6f681031d5f24, references: 46, referenceMentions: 44
SUMMARY: cff38587ec97f02aed4c04a7abf2cd185c4cfda5, references: 23, referenceMentions: 38
SUMMARY: e4faf2c1d76b9bf8f8b4524dfb8c5c6b93be5f35, references: 0, referenceMentions: 0
joannap@joannap:/ws/projects/medical-citations$ 
joannap@joannap:/ws/projects/medical-citations$ grep SUMMARY b313ca7/paper-data.b313ca7.txt 
SUMMARY: 0c5cf03abbcaeb86585ef110141b76ef8765ceb9, references: 19, referenceMentions: 23
SUMMARY: 12cd18f35b83d3559c2be283d1ab0d513bac3ed9, references: 19, referenceMentions: 22
SUMMARY: 1925673c0ac47e9024440db9133d5214fc4104a6, references: 15, referenceMentions: 0
SUMMARY: 1dca9541f4b0fc54c68bd58980602f2dc16662d9, references: 1, referenceMentions: 1
SUMMARY: 295a641761198ea565fe7e877912677f61680dc4, references: 47, referenceMentions: 46
SUMMARY: 2cd5827927d5f960b22cd0db843ae4a56953fb89, references: 85, referenceMentions: 106
SUMMARY: 58d74b5812ce9c975145f592848d3ae6204082af, references: 0, referenceMentions: 0
SUMMARY: 676d59b9330e084b2d5a3718233396a8bd2d02a4, references: 41, referenceMentions: 0
SUMMARY: 68a88ae2b8c586a49876ff4d0af871ed72d42b66, references: 53, referenceMentions: 47
SUMMARY: 770509379ba6d456d35734e8dba0c6367e6e26c2, references: 56, referenceMentions: 62
SUMMARY: 7e9537bbead3c32c07496f7feb787677a2e77613, references: 52, referenceMentions: 65
SUMMARY: 830476f35dc4ad59f86ece5028a030b05ab83696, references: 25, referenceMentions: 32
SUMMARY: 9444ba92b6d5d431e7bb3be26a2d6133f38c5485, references: 26, referenceMentions: 25
SUMMARY: 9e257dad7923d5b44b236ea03fd2c9d654c06a25, references: 30, referenceMentions: 30
SUMMARY: 9f77b3c5b0b821acbc45098e036c3cdc306a4716, references: 25, referenceMentions: 38
SUMMARY: b48ab369f5a3515b9b347cff89b713af51b06866, references: 39, referenceMentions: 61
SUMMARY: b5252acfddf1868c6b1a8078d6c6f681031d5f24, references: 46, referenceMentions: 44
SUMMARY: cff38587ec97f02aed4c04a7abf2cd185c4cfda5, references: 23, referenceMentions: 38
SUMMARY: e4faf2c1d76b9bf8f8b4524dfb8c5c6b93be5f35, references: 58, referenceMentions: 74
```

One "good" reference is lost, but upon inspection it is clear that the lost reference isn't correct even in the gold data. With these changes, references and reference mentions are found for 6 of the 7 papers that previously had neither.

**Comment by jpowerwa** [comment]
Attached are contents of references and reference mentions for both master and this set of changes.
[paper-data.gold.txt](https://github.com/allenai/science-parse/files/2294298/paper-data.gold.txt)
[paper-data.b313ca7.txt](https://github.com/allenai/science-parse/files/2294300/paper-data.b313ca7.txt)

**Comment by jpowerwa** [comment]
Also fixes some incorrect reference dates and extraneous spaces in the content of existing references. The incorrect dates were being extracted from the journal information in the header of the paper, which sometimes is pulled into the reference information when the references span multiple pages. 

Attached is a diff of the per-paper changes for papers that had references already.
[per-paper-diff.b313ca7.txt](https://github.com/allenai/science-parse/files/2294424/per-paper-diff.b313ca7.txt)

**Comment by jpowerwa** [comment]
```
joannap@joannap:/ws/git/science-parse$ sbt
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
[info] Loading project definition from /ws/git/science-parse/project
[info] Set current project to science-parse (in build file:/ws/git/science-parse/)
> test

...

===============================================
Tests
Total tests run: 26, Failures: 0, Skips: 0
===============================================

[info] ScalaTest
[info] Run completed in 6 minutes, 25 seconds.
[info] Total number of tests run: 35
[info] Suites: completed 5, aborted 0
[info] Tests: succeeded 35, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Passed: Total 35, Failed 0, Errors 0, Passed 35
[success] Total time: 392 s, completed Aug 16, 2018 9:37:55 AM
```

File: core/src/main/java/org/allenai/scienceparse/PDFDocToPartitionedText.java [hunk 1]
```diff
@@ -339,6 +340,13 @@ private static boolean firstCol(PDFLine line, PDFPage p) {
     return PDFToCRFInput.getX(line, true) < pageThird;
   }
 
+  private static boolean referenceIsSplit(String lineOne, String lineTwo) {
+    Matcher lineOneMatcher = referenceStartPattern.matcher(lineOne);
+    return (lineOneMatcher.find() &&
+            lineOneMatcher.end() == lineOne.length() &&
+            !referenceStartPattern.matcher(lineTwo).find());
```
Comment (dirkgr) [inline core/src/main/java/org/allenai/scienceparse/PDFDocToPartitionedText.java]: This should really use `RegexWithTimeout`, for safety.

**Comment by dirkgr** [comment]
Woha. A blast from the past. Why did you resurrect this?

**Comment by dirkgr** [comment]
Maybe the real question is, why didn't you merge it last year?

**Pull request merged**